Bug: Incosistent Metadata after updating the Online Key

### What happened?

The Online Key is used for signing the Targets, Snapshop, Timestamp, and the Target Delegated Roles (succinct roles `bins-<id>.json`)

The bug occurs when the Online Key is updated, generating an inconsistency while the TUF client tries to download a Target File (artifact).

After an investigation, the root cause of the bug is that we bump the high-level Target roles but don't update the keys, in this case, the Online Key in the Target Metadata (`targets.signed.delegations.keys` and `targets.signed.delegations.succinct_roles.keyids`).

The below test shows the failure and the inconsistency.
I modified our Functional Tests for Metadata Update to rotate the online key.

```logs
INFO     tests.functional.metadata.test_update:test_update.py:22 Adding artifacts
INFO     tests.functional.metadata.test_update:test_update.py:45 Added task_id: d213ca176460459aa0a2c58f32858d1e
INFO     tests.functional.metadata.test_update:test_update.py:22 Adding artifacts
INFO     tests.functional.metadata.test_update:test_update.py:45 Added task_id: 4c974c106e054d758bae664749b6923b
INFO     tests.functional.metadata.test_update:test_update.py:22 Adding artifacts
INFO     tests.functional.metadata.test_update:test_update.py:45 Added task_id: f2050b62642d4be8aeb75b755ae83405
INFO     tests.functional.metadata.test_update:test_update.py:22 Adding artifacts
INFO     tests.functional.metadata.test_update:test_update.py:45 Added task_id: d3a171d70e7f4d53a1155ab3afc9a902
INFO     tests.functional.metadata.test_update:test_update.py:90 [METADATA UPDATE] Submiting Root Metadata Update
INFO     tests.functional.metadata.test_update:test_update.py:116 [METADATA UPDATE]  Metadata Updated by 32ceccb5382e4010a96dda9abad4b383
INFO     tests.functional.metadata.test_update:test_update.py:22 Adding artifacts
INFO     tests.functional.metadata.test_update:test_update.py:45 Added task_id: 7330cc86307c4f64a87ee7b643363185
INFO     tests.functional.metadata.test_update:test_update.py:146 [METADATA UPDATE] {"data":{"task_id":"32ceccb5382e4010a96dda9abad4b383","state":"SUCCESS","result":{"message":"Metadata Update Processed","status":true,"task":"metadata_update","last_update":"2024-04-17T12:05:33.071964Z","details":{"role":"root"}}},"message":"Task state."}
INFO     tests.functional.metadata.test_update:test_update.py:153 [METADATA UPDATE] Update Metadata to 2.root.json finished
INFO     tests.functional.metadata.test_update:test_update.py:22 Adding artifacts
INFO     tests.functional.metadata.test_update:test_update.py:45 Added task_id: 467cd789f552414ca21734de9702dd99
INFO     tests.functional.metadata.test_update:test_update.py:22 Adding artifacts
INFO     tests.functional.metadata.test_update:test_update.py:45 Added task_id: ac1a7ee237844acfbe861975fae02e9e
INFO     tests.functional.metadata.test_update:test_update.py:22 Adding artifacts
INFO     tests.functional.metadata.test_update:test_update.py:45 Added task_id: bb7aff18b7a14d6a9243664bd3b5e39e
INFO     tests.functional.metadata.test_update:test_update.py:22 Adding artifacts
INFO     tests.functional.metadata.test_update:test_update.py:45 Added task_id: 68a28dcbd4e94dc184d3e3d4440348d8
INFO     tests.functional.metadata.test_update:test_update.py:175 [METADATA UPDATE] Metadata Update available (2.root.json)
INFO     tests.functional.metadata.test_update:test_update.py:49 Stop adding artifacts. Total requests: 9
INFO     tests.functional.metadata.test_update:test_update.py:188 Task 1/9 finshed!
INFO     tests.functional.metadata.test_update:test_update.py:188 Task 2/9 finshed!
INFO     tests.functional.metadata.test_update:test_update.py:188 Task 3/9 finshed!
INFO     tests.functional.metadata.test_update:test_update.py:188 Task 4/9 finshed!
INFO     tests.functional.metadata.test_update:test_update.py:188 Task 5/9 finshed!
INFO     tests.functional.metadata.test_update:test_update.py:188 Task 6/9 finshed!
INFO     tests.functional.metadata.test_update:test_update.py:188 Task 7/9 finshed!
INFO     tests.functional.metadata.test_update:test_update.py:188 Task 8/9 finshed!
INFO     tests.functional.metadata.test_update:test_update.py:188 Task 9/9 finshed!
INFO     tests.functional.metadata.test_update:test_update.py:195 Verifying test/condescending_golick-0.tar.gz
============================================================================= warnings summary ==============================================================================
tests/functional/metadata/test_update.py::test_updating_root_metadata_full_signed
tests/functional/metadata/test_update.py::test_updating_root_metadata_full_signed
tests/functional/metadata/test_update.py::test_updating_root_metadata_full_signed
  /usr/local/lib/python3.10/site-packages/pytest_bdd/compat.py:46: PytestDeprecationWarning: A private pytest class or function was used.
    fd = FixtureDef(

tests/functional/metadata/test_update.py::test_updating_root_metadata_full_signed
tests/functional/metadata/test_update.py::test_updating_root_metadata_full_signed
tests/functional/metadata/test_update.py::test_updating_root_metadata_full_signed
  /usr/local/lib/python3.10/site-packages/pytest_html/basereport.py:356: DeprecationWarning: The 'py' module is deprecated and support will be removed in a future release.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
---------------------------------------------------- generated json file: /rstuf-runner/rstuf-umbrella/test-report.json -----------------------------------------------------
============================================================================= slowest durations =============================================================================
16.34s call     tests/functional/metadata/test_update.py::test_updating_root_metadata_full_signed
------------------------------------------------ Generated html report: file:///rstuf-runner/rstuf-umbrella/test-report.html ------------------------------------------------
========================================================================== short test summary info ==========================================================================
FAILED tests/functional/metadata/test_update.py::test_updating_root_metadata_full_signed - tuf.api.exceptions.UnsignedMetadataError: bins-1 was signed by 0/1 keys
====================================================================== 1 failed, 6 warnings in 16.42s =======================================================================
```


Below, I added a simple fix that shows the consistency.

```logs

tests/functional/metadata/test_update.py::test_updating_root_metadata_full_signed <- ../../usr/local/lib/python3.10/site-packages/pytest_bdd/scenario.py 
------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------
2024-04-17 11:58:37 [    INFO] Adding artifacts (test_update.py:22)
2024-04-17 11:58:37 [    INFO] Added task_id: 5791656a48694886b80d4f8540d3df8d (test_update.py:45)
2024-04-17 11:58:38 [    INFO] Adding artifacts (test_update.py:22)
2024-04-17 11:58:38 [    INFO] Added task_id: 9c3c85acf42b4713b742162130799da1 (test_update.py:45)
2024-04-17 11:58:38 [    INFO] Adding artifacts (test_update.py:22)
2024-04-17 11:58:38 [    INFO] Added task_id: d6331645182e417eb5f51aa60276e8bd (test_update.py:45)
2024-04-17 11:58:39 [    INFO] Adding artifacts (test_update.py:22)
2024-04-17 11:58:39 [    INFO] Added task_id: 79806abceb95462fa42f4056e334a498 (test_update.py:45)
2024-04-17 11:58:39 [    INFO] [METADATA UPDATE] Submiting Root Metadata Update (test_update.py:90)
2024-04-17 11:58:39 [    INFO] [METADATA UPDATE]  Metadata Updated by ed2a1afb04414feb8e39fc876d1e3ae1 (test_update.py:116)
2024-04-17 11:58:40 [    INFO] Adding artifacts (test_update.py:22)
2024-04-17 11:58:40 [    INFO] Added task_id: 64c6df38a2b2421f8f9b9afe71901f0b (test_update.py:45)
2024-04-17 11:58:40 [    INFO] [METADATA UPDATE] {"data":{"task_id":"ed2a1afb04414feb8e39fc876d1e3ae1","state":"SUCCESS","result":{"message":"Metadata Update Processed","status":true,"task":"metadata_update","last_update":"2024-04-17T11:58:40.200730Z","details":{"role":"root"}}},"message":"Task state."} (test_update.py:146)
2024-04-17 11:58:40 [    INFO] [METADATA UPDATE] Update Metadata to 2.root.json finished (test_update.py:153)
2024-04-17 11:58:40 [    INFO] Adding artifacts (test_update.py:22)
2024-04-17 11:58:40 [    INFO] Added task_id: 87d6476a0dbd43b4b3562cceac8241af (test_update.py:45)
2024-04-17 11:58:41 [    INFO] Adding artifacts (test_update.py:22)
2024-04-17 11:58:41 [    INFO] Added task_id: 5859bac4faa5404cb730fb4e139c9d16 (test_update.py:45)
2024-04-17 11:58:41 [    INFO] Adding artifacts (test_update.py:22)
2024-04-17 11:58:41 [    INFO] Added task_id: 40a514a128ef49d9b40ad8a37be9736f (test_update.py:45)
2024-04-17 11:58:42 [    INFO] Adding artifacts (test_update.py:22)
2024-04-17 11:58:42 [    INFO] Added task_id: 48a7d7ca1c644657bf848d4685147204 (test_update.py:45)
2024-04-17 11:58:42 [    INFO] [METADATA UPDATE] Metadata Update available (2.root.json) (test_update.py:175)
2024-04-17 11:58:42 [    INFO] Stop adding artifacts. Total requests: 9 (test_update.py:49)
2024-04-17 11:58:46 [    INFO] Task 1/9 finshed! (test_update.py:188)
2024-04-17 11:58:47 [    INFO] Task 2/9 finshed! (test_update.py:188)
2024-04-17 11:58:51 [    INFO] Task 3/9 finshed! (test_update.py:188)
2024-04-17 11:58:51 [    INFO] Task 4/9 finshed! (test_update.py:188)
2024-04-17 11:58:52 [    INFO] Task 5/9 finshed! (test_update.py:188)
2024-04-17 11:58:52 [    INFO] Task 6/9 finshed! (test_update.py:188)
2024-04-17 11:58:52 [    INFO] Task 7/9 finshed! (test_update.py:188)
2024-04-17 11:58:52 [    INFO] Task 8/9 finshed! (test_update.py:188)
2024-04-17 11:58:52 [    INFO] Task 9/9 finshed! (test_update.py:188)
2024-04-17 11:58:52 [    INFO] Verifying test/great_kapitsa-0.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/serene_blackburn-1.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/amazing_edison-2.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/adoring_tu-3.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/optimistic_chatterjee-4.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/amazing_visvesvaraya-5.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/optimistic_gould-6.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/lucid_dijkstra-7.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/stoic_gates-8.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/mystifying_colden-9.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/practical_brattain-0.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/pedantic_dirac-1.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/modest_joliot-2.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/youthful_dirac-3.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/amazing_khayyam-4.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/vigorous_wu-5.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/nostalgic_jepsen-6.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/dazzling_shirley-7.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/affectionate_edison-8.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/dazzling_ellis-9.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/sweet_colden-0.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/vigilant_rosalind-1.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/busy_saha-2.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/elated_leavitt-3.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/goofy_herschel-4.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/affectionate_brattain-5.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/tender_wiles-6.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/hardcore_austin-7.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/cranky_jang-8.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/ecstatic_faraday-9.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/beautiful_euler-0.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/compassionate_johnson-1.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/loving_chaplygin-2.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/gracious_feistel-3.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/xenodochial_buck-4.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/bold_haibt-5.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/pedantic_wright-6.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/quizzical_haslett-7.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/kind_heyrovsky-8.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/keen_ritchie-9.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/sad_johnson-0.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/keen_feistel-1.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/ecstatic_cannon-2.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/xenodochial_haslett-3.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/zen_wu-4.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/upbeat_jackson-5.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/boring_ptolemy-6.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/amazing_darwin-7.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/elastic_cartwright-8.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/inspiring_dubinsky-9.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/strange_snyder-0.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/nostalgic_wescoff-1.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/eloquent_jang-2.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/quizzical_bhabha-3.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/competent_hamilton-4.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/vigorous_mccarthy-5.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/pensive_banach-6.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/charming_bohr-7.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/lucid_lalande-8.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/zealous_chatelet-9.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/awesome_rosalind-0.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/youthful_dijkstra-1.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/modest_merkle-2.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/elated_noyce-3.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/sad_carson-4.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/quirky_leavitt-5.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/naughty_varahamihira-6.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/awesome_darwin-7.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/recursing_lovelace-8.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/stupefied_kalam-9.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/kind_solomon-0.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/naughty_colden-1.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/suspicious_darwin-2.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/sleepy_golick-3.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/brave_nash-4.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/elegant_benz-5.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/condescending_brown-6.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/optimistic_gould-7.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/boring_vaughan-8.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/suspicious_torvalds-9.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/condescending_mendeleev-0.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/elastic_bassi-1.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/boring_napier-2.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/elegant_sutherland-3.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/zealous_shaw-4.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/ecstatic_chatterjee-5.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/wonderful_feynman-6.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/xenodochial_dubinsky-7.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/inspiring_robinson-8.tar.gz (test_update.py:195)
2024-04-17 11:58:52 [    INFO] Verifying test/fervent_ride-9.tar.gz (test_update.py:195)


Feature: Update metadata
    Scenario: Updating Root metadata full signed
        Given RSTUF is running and operational
        Then the RSTUF is receiving multiple requests
        When the RSTUF key holders send a fully signed metadata
        Then the API requester should get status code '202' with 'task_id'
        Then the API requester gets from endpoint 'GET /api/v1/task' status 'SUCCESS'
        Then the '2.root.json' will be available in the TUF Metadata
        Then the user downloads will not have inconsistency during this process
    PASSED

```

Credit for the bug: @matglas 


Note: Another bug was identified during the investigation.

* The root persistency in the storage should be done in the end together with the timestamp during the Metadata Update

https://github.com/repository-service-tuf/repository-service-tuf-worker/blob/70f548f951850f84a4e0677ad87483494533b65f/repository_service_tuf_worker/repository.py#L1304-L1311

### What steps did you take?

1. Generate a Ceremony `rstuf admin ceremony`
2. Add targets
3. Generate a Metadata Update and rotate the online key `rstuf admin metadata update`
4. Add targets
5. Try to download targets using a TUF client

### What behavior did you expect?

Update the Online Key without breaking the TUF Metadata consistency for the TUF clients.

Include a Functional Test to avoid regression

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/repository_service_tuf_worker/repository.py
+++ b/repository_service_tuf_worker/repository.py
@@ -138,7 +138,11 @@ class MetadataRepository:
             # All roles except root share the same one key and it doesn't
             # matter from which role we will get the key.
             keyid: str = root.signed.roles["timestamp"].keyids[0]
-            return root.signed.keys[keyid]
+            key = root.signed.keys[keyid]
+            key_dict = key.to_dict()
+            key_dict["keyid"] = key.keyid
+            self.write_repository_settings("ONLINE_KEY", key_dict)
+            return key
 
     @_online_key.setter
     def _online_key(self, key: Key):

--- a/repository_service_tuf_worker/repository.py
+++ b/repository_service_tuf_worker/repository.py
@@ -115,7 +115,6 @@ class MetadataRepository:
         app_settings = self.refresh_settings()
         self._storage_backend: IStorage = app_settings.STORAGE
         self._signer_store = SignerStore(app_settings)
-        self._online_key: Key
         self._db = app_settings.SQL
         self._redis = redis.StrictRedis.from_url(
             self._worker_settings.REDIS_SERVER

--- a/repository_service_tuf_worker/repository.py
+++ b/repository_service_tuf_worker/repository.py
@@ -378,6 +378,30 @@ class MetadataRepository:
 
         return snapshot.signed.version
 
+    def _update_targets_delegations_key(self, targets: Metadata[Targets]):
+        """
+        Update the key used by delegations referenced in targets metadata.
+
+        Args:
+            targets: current top level targets metadata.
+        """
+        current_root = self._storage_backend.get(Root.type)
+        old_online_keyid = current_root.signed.roles[Targets.type].keyids[0]
+        # New online keyid is set to self._online_key before function call.
+        new_online_key = self._online_key
+        if old_online_keyid == new_online_key.keyid:
+            return
+
+        bins = True if targets.signed.delegations.succinct_roles else False
+        if bins:
+            targets.signed.revoke_key(old_online_keyid)
+            targets.signed.add_key(new_online_key)
+
+        else:
+            for role in targets.signed.delegations.roles.values():
+                targets.signed.revoke_key(old_online_keyid, role.name)
+                targets.signed.add_key(new_online_key, role.name)
+
     def _get_role_for_artifact_path(self, artifact_path: str) -> Optional[str]:
         """
         Return role name by target file path
@@ -1095,8 +1119,9 @@ class MetadataRepository:
             if force or (targets.signed.expires - today) < timedelta(
                 hours=self._hours_before_expire
             ):
-                logging.info("Bumped version of 'Targets' role")
+                self._update_targets_delegations_key(targets)
                 self._bump_and_persist(targets, Targets.type)
+                logging.info("Bumped version of 'Targets' role")
                 snapshot_bump = True
 
         if force:

--- a/repository_service_tuf_worker/repository.py
+++ b/repository_service_tuf_worker/repository.py
@@ -115,6 +115,7 @@ class MetadataRepository:
         app_settings = self.refresh_settings()
         self._storage_backend: IStorage = app_settings.STORAGE
         self._signer_store = SignerStore(app_settings)
+        self._online_key: Key
         self._db = app_settings.SQL
         self._redis = redis.StrictRedis.from_url(
             self._worker_settings.REDIS_SERVER
@@ -203,14 +204,8 @@ class MetadataRepository:
         The metadata role type is used as default key id. This is only allowed
         for top-level roles.
         """
-        role.signatures.clear()
-        root: Metadata[Root] = self._storage_backend.get("root")
-        # All roles except root share the same one key and it doesn't matter
-        # from which role we will get the key.
-        keyid: str = root.signed.roles["timestamp"].keyids[0]
-        public_key = root.signed.keys[keyid]
-        signer = self._signer_store.get(public_key)
-        role.sign(signer, append=True)
+        signer = self._signer_store.get(self._online_key)
+        role.sign(signer)
 
     def _persist(self, role: Metadata, role_name: str) -> str:
         """
@@ -664,9 +659,13 @@ class MetadataRepository:
         custom_delegated_roles: Dict[str, Any] = self._settings.get_fresh(
             "CUSTOM_DELEGATED_ROLES"
         )
-        self._persist(root, Root.type)
-        self.write_repository_settings("ROOT_SIGNING", None)
+        # All roles except root share the same one key and it doesn't matter
+        # from which role we will get the key.
+        keyid: str = root.signed.roles["timestamp"].keyids[0]
+        self._online_key = root.signed.keys[keyid]
         self._bootstrap_online_roles(root, custom_delegated_roles)
+        self.write_repository_settings("ROOT_SIGNING", None)
+        self._persist(root, Root.type)
         self.write_repository_settings("BOOTSTRAP", task_id)
 
     def bootstrap(
@@ -1424,6 +1423,20 @@ class MetadataRepository:
             status_lock_targets = False
             try:
                 with self._redis.lock(LOCK_TARGETS, timeout=self._timeout):
+                    curr_online_key = self._online_key
+                    # All roles except root share the same one key and it
+                    # doesn't matter from which role we will get the key.
+                    keyid: str = new_root.signed.roles["timestamp"].keyids[0]
+                    self._online_key = new_root.signed.keys[keyid]
+
+                    try:
+                        self._run_online_roles_bump(force=True)
+                        logging.info("Updating all targets metadata")
+                    except Exception as e:
+                        # Recover to correct online key state.
+                        self._online_key = curr_online_key
+                        raise e
+
                     # root metadata and online key are updated
                     # 1. persist the new root
                     # 2. bump all target roles
@@ -1431,8 +1444,6 @@ class MetadataRepository:
                     logging.info(
                         f"Updating root metadata: {new_root.signed.version}"
                     )
-                    self._run_online_roles_bump(force=True)
-                    logging.info("Updating all targets metadata")
                     status_lock_targets = True
             except redis.exceptions.LockNotOwnedError:
                 if status_lock_targets is False:

--- a/repository_service_tuf_worker/repository.py
+++ b/repository_service_tuf_worker/repository.py
@@ -1330,7 +1330,7 @@ class MetadataRepository:
             },
         )
 
-    def _trusted_root_update(
+    def _verify_new_root_signing(
         self, current_root: Metadata[Root], new_root: Metadata[Root]
     ):
         """Verify if the new metadata is a trusted Root metadata"""
@@ -1361,7 +1361,7 @@ class MetadataRepository:
         current_root: Metadata[Root] = self._storage_backend.get(Root.type)
 
         try:
-            self._trusted_root_update(current_root, new_root)
+            self._verify_new_root_signing(current_root, new_root)
 
         except UnsignedMetadataError:
             # TODO: Add missing sanity check - new root must have at least 1

--- a/repository_service_tuf_worker/repository.py
+++ b/repository_service_tuf_worker/repository.py
@@ -129,6 +129,24 @@ class MetadataRepository:
     def _settings(self) -> Dynaconf:
         return get_repository_settings()
 
+    @property
+    def _online_key(self) -> Key:
+        key_dict = self._settings.get_fresh("ONLINE_KEY")
+        if key_dict is not None:
+            return Key.from_dict(key_dict.pop("keyid"), key_dict)
+        else:
+            root: Metadata[Root] = self._storage_backend.get(Root.type)
+            # All roles except root share the same one key and it doesn't
+            # matter from which role we will get the key.
+            keyid: str = root.signed.roles["timestamp"].keyids[0]
+            return root.signed.keys[keyid]
+
+    @_online_key.setter
+    def _online_key(self, key: Key):
+        key_dict = key.to_dict()
+        key_dict["keyid"] = key.keyid
+        self.write_repository_settings("ONLINE_KEY", key_dict)
+
     @classmethod
     def create_service(cls) -> "MetadataRepository":
         """Class Method for MetadataRepository service creation."""
@@ -1432,9 +1450,9 @@ class MetadataRepository:
         # We always persist the new root metadata, but we cannot persist
         # without verifying if the online key is rotated to avoid a mismatch
         # with the rest of the roles using the online key.
-        current_online_keyid = current_root.signed.roles[Timestamp.type].keyids
-        new_online_keyid = new_root.signed.roles[Timestamp.type].keyids
-        if current_online_keyid == new_online_keyid:
+        current_keyid = current_root.signed.roles[Timestamp.type].keyids[0]
+        new_keyid = new_root.signed.roles[Timestamp.type].keyids[0]
+        if current_keyid == new_keyid:
             # online key is not changed, persist the new root
             self._persist(new_root, Root.type)
             logging.info(f"Updating root metadata: {new_root.signed.version}")
@@ -1449,10 +1467,7 @@ class MetadataRepository:
             try:
                 with self._redis.lock(LOCK_TARGETS, timeout=self._timeout):
                     curr_online_key = self._online_key
-                    # All roles except root share the same one key and it
-                    # doesn't matter from which role we will get the key.
-                    keyid: str = new_root.signed.roles["timestamp"].keyids[0]
-                    self._online_key = new_root.signed.keys[keyid]
+                    self._online_key = new_root.signed.keys[new_keyid]
 
                     try:
                         self._run_online_roles_bump(force=True)

--- a/tests/unit/tuf_repository_service_worker/test_repository.py
+++ b/tests/unit/tuf_repository_service_worker/test_repository.py
@@ -111,7 +111,11 @@ class TestMetadataRepository:
             "get_repository_settings",
             lambda *a, **kw: fake_settings,
         )
-        key = pretend.stub(keyid="key_id")
+        key_dict = {"keyval": "old_bar"}
+        expected_dict = copy(key_dict)
+        key = pretend.stub(
+            keyid="key_id", to_dict=pretend.call_recorder(lambda: key_dict)
+        )
         fake_root = pretend.stub(
             signed=pretend.stub(
                 roles={"timestamp": pretend.stub(keyids=["key_id"])},
@@ -122,11 +126,19 @@ class TestMetadataRepository:
         test_repo._storage_backend.get = pretend.call_recorder(
             lambda *a: fake_root
         )
+        test_repo.write_repository_settings = pretend.call_recorder(
+            lambda *a: None
+        )
         result = test_repo._online_key
         assert result == key
         assert fake_settings.get_fresh.calls == [pretend.call("ONLINE_KEY")]
         assert test_repo._storage_backend.get.calls == [
             pretend.call(Root.type)
+        ]
+        expected_dict["keyid"] = key.keyid
+        assert key.to_dict.calls == [pretend.call()]
+        assert test_repo.write_repository_settings.calls == [
+            pretend.call("ONLINE_KEY", expected_dict)
         ]
 
     def test_refresh_settings_with_none_arg(self, test_repo):

--- a/tests/unit/tuf_repository_service_worker/test_repository.py
+++ b/tests/unit/tuf_repository_service_worker/test_repository.py
@@ -5,6 +5,7 @@
 
 import datetime
 from contextlib import contextmanager
+from copy import copy
 from datetime import timezone
 from math import log
 
@@ -24,6 +25,8 @@ from tuf.api.metadata import (
 
 from repository_service_tuf_worker import Dynaconf, repository
 from repository_service_tuf_worker.models import targets_schema
+
+REPOSITORY_PATH = f"repository_service_tuf_worker.repository"
 
 
 class TestRoles:
@@ -76,6 +79,56 @@ class TestMetadataRepository:
         result = test_repo._settings
 
         assert result == fake_settings
+
+    def test_online_key_property_from_redis(self, test_repo, monkeypatch):
+        fake_key_dict = {"keyval": "foo", "keyid": "key_id"}
+        fake_settings = pretend.stub(
+            get_fresh=pretend.call_recorder(lambda a: copy(fake_key_dict))
+        )
+        monkeypatch.setattr(
+            repository,
+            "get_repository_settings",
+            lambda *a, **kw: fake_settings,
+        )
+        key = pretend.stub(keyid="key_id")
+        fake_key_obj = pretend.stub(
+            from_dict=pretend.call_recorder(lambda *a: key)
+        )
+        monkeypatch.setattr(f"{REPOSITORY_PATH}.Key", fake_key_obj)
+        result = test_repo._online_key
+        assert result == key
+        assert fake_settings.get_fresh.calls == [pretend.call("ONLINE_KEY")]
+        assert fake_key_obj.from_dict.calls == [
+            pretend.call(fake_key_dict.pop("keyid"), fake_key_dict)
+        ]
+
+    def test_online_key_property_from_storage(self, test_repo, monkeypatch):
+        fake_key_dict = {"keyval": "foo", "keyid": "key_id"}
+        fake_settings = pretend.stub(
+            get_fresh=pretend.call_recorder(lambda a: None)
+        )
+        monkeypatch.setattr(
+            repository,
+            "get_repository_settings",
+            lambda *a, **kw: fake_settings,
+        )
+        key = pretend.stub(keyid="key_id")
+        fake_root = pretend.stub(
+            signed=pretend.stub(
+                roles={"timestamp": pretend.stub(keyids=["key_id"])},
+                keys={"key_id": key},
+                version=2,
+            )
+        )
+        test_repo._storage_backend.get = pretend.call_recorder(
+            lambda *a: fake_root
+        )
+        result = test_repo._online_key
+        assert result == key
+        assert fake_settings.get_fresh.calls == [pretend.call("ONLINE_KEY")]
+        assert test_repo._storage_backend.get.calls == [
+            pretend.call(Root.type)
+        ]
 
     def test_refresh_settings_with_none_arg(self, test_repo):
         test_repo.refresh_settings()
@@ -190,8 +243,21 @@ class TestMetadataRepository:
         assert "No permission /run/secrets/*" in str(e)
         assert "No permission /run/secrets/*" == caplog.messages[0]
 
-    def test__sign(self, test_repo):
-        test_repo._online_key = "onlin_key"
+    def test__sign(self, test_repo, monkeypatch):
+        fake_key_dict = {"keyval": "foo", "keyid": "keyid"}
+        fake_settings = pretend.stub(
+            get_fresh=pretend.call_recorder(lambda a: copy(fake_key_dict))
+        )
+        monkeypatch.setattr(
+            repository,
+            "get_repository_settings",
+            lambda *a, **kw: fake_settings,
+        )
+        fake_key = pretend.stub(keyid="new_key")
+        fake_key_obj = pretend.stub(
+            from_dict=pretend.call_recorder(lambda *a: fake_key)
+        )
+        monkeypatch.setattr(f"{REPOSITORY_PATH}.Key", fake_key_obj)
         fake_md = pretend.stub(
             sign=pretend.call_recorder(lambda *a, **kw: None),
         )
@@ -202,7 +268,11 @@ class TestMetadataRepository:
         test_result = test_repo._sign(fake_md)
 
         assert test_result is None
-        assert test_repo._signer_store.get.calls == [pretend.call("onlin_key")]
+        assert fake_settings.get_fresh.calls == [pretend.call("ONLINE_KEY")]
+        assert fake_key_obj.from_dict.calls == [
+            pretend.call(fake_key_dict.pop("keyid"), fake_key_dict)
+        ]
+        assert test_repo._signer_store.get.calls == [pretend.call(fake_key)]
         assert fake_md.sign.calls == [pretend.call("key_signer_1")]
 
     def _test_helper_persist(
@@ -871,9 +941,23 @@ class TestMetadataRepository:
             pretend.call(version=mocked_targets.signed.version),
         ]
 
-    def test__update_targets_delegations_key_bins(self, test_repo):
-        fake_online_key_stub = pretend.stub(keyid="new_online_key")
-        test_repo._online_key = fake_online_key_stub
+    def test__update_targets_delegations_key_bins(
+        self, test_repo, monkeypatch
+    ):
+        fake_key_dict = {"keyval": "foo", "keyid": "keyid"}
+        fake_settings = pretend.stub(
+            get_fresh=pretend.call_recorder(lambda a: copy(fake_key_dict))
+        )
+        monkeypatch.setattr(
+            repository,
+            "get_repository_settings",
+            lambda *a, **kw: fake_settings,
+        )
+        fake_key = pretend.stub(keyid="new_key")
+        fake_key_obj = pretend.stub(
+            from_dict=pretend.call_recorder(lambda *a: fake_key)
+        )
+        monkeypatch.setattr(f"{REPOSITORY_PATH}.Key", fake_key_obj)
         fake_root = pretend.stub(
             signed=pretend.stub(
                 roles={Targets.type: pretend.stub(keyids=["old_keyid"])}
@@ -895,18 +979,32 @@ class TestMetadataRepository:
         assert test_repo._storage_backend.get.calls == [
             pretend.call(Root.type)
         ]
+        assert fake_settings.get_fresh.calls == [pretend.call("ONLINE_KEY")]
+        assert fake_key_obj.from_dict.calls == [
+            pretend.call(fake_key_dict.pop("keyid"), fake_key_dict)
+        ]
         assert fake_targets.signed.revoke_key.calls == [
             pretend.call("old_keyid")
         ]
-        assert fake_targets.signed.add_key.calls == [
-            pretend.call(fake_online_key_stub)
-        ]
+        assert fake_targets.signed.add_key.calls == [pretend.call(fake_key)]
 
     def test__update_targets_delegations_key_custom_delegations(
-        self, test_repo
+        self, test_repo, monkeypatch
     ):
-        fake_online_key_stub = pretend.stub(keyid="new_online_key")
-        test_repo._online_key = fake_online_key_stub
+        fake_key_dict = {"keyval": "foo", "keyid": "keyid"}
+        fake_settings = pretend.stub(
+            get_fresh=pretend.call_recorder(lambda a: copy(fake_key_dict))
+        )
+        monkeypatch.setattr(
+            repository,
+            "get_repository_settings",
+            lambda *a, **kw: fake_settings,
+        )
+        fake_key = pretend.stub(keyid="new_key")
+        fake_key_obj = pretend.stub(
+            from_dict=pretend.call_recorder(lambda *a: fake_key)
+        )
+        monkeypatch.setattr(f"{REPOSITORY_PATH}.Key", fake_key_obj)
         fake_root = pretend.stub(
             signed=pretend.stub(
                 roles={Targets.type: pretend.stub(keyids=["old_keyid"])}
@@ -937,6 +1035,10 @@ class TestMetadataRepository:
         assert test_repo._storage_backend.get.calls == [
             pretend.call(Root.type)
         ]
+        assert fake_settings.get_fresh.calls == [pretend.call("ONLINE_KEY")]
+        assert fake_key_obj.from_dict.calls == [
+            pretend.call(fake_key_dict.pop("keyid"), fake_key_dict)
+        ]
         assert fake_targets.signed.delegations.roles.values.calls == [
             pretend.call(),
         ]
@@ -945,14 +1047,27 @@ class TestMetadataRepository:
             pretend.call("old_keyid", "role2"),
         ]
         assert fake_targets.signed.add_key.calls == [
-            pretend.call(fake_online_key_stub, "role1"),
-            pretend.call(fake_online_key_stub, "role2"),
+            pretend.call(fake_key, "role1"),
+            pretend.call(fake_key, "role2"),
         ]
 
     def test__update_targets_delegations_online_key_not_changed(
-        self, test_repo
+        self, test_repo, monkeypatch
     ):
-        test_repo._online_key = pretend.stub(keyid="online_key")
+        fake_key_dict = {"keyval": "foo", "keyid": "keyid"}
+        fake_settings = pretend.stub(
+            get_fresh=pretend.call_recorder(lambda a: copy(fake_key_dict))
+        )
+        monkeypatch.setattr(
+            repository,
+            "get_repository_settings",
+            lambda *a, **kw: fake_settings,
+        )
+        fake_key = pretend.stub(keyid="online_key")
+        fake_key_obj = pretend.stub(
+            from_dict=pretend.call_recorder(lambda *a: fake_key)
+        )
+        monkeypatch.setattr(f"{REPOSITORY_PATH}.Key", fake_key_obj)
         fake_root = pretend.stub(
             signed=pretend.stub(
                 roles={Targets.type: pretend.stub(keyids=["online_key"])}
@@ -966,6 +1081,10 @@ class TestMetadataRepository:
         assert result is None
         assert test_repo._storage_backend.get.calls == [
             pretend.call(Root.type)
+        ]
+        assert fake_settings.get_fresh.calls == [pretend.call("ONLINE_KEY")]
+        assert fake_key_obj.from_dict.calls == [
+            pretend.call(fake_key_dict.pop("keyid"), fake_key_dict)
         ]
 
     def test__get_role_for_artifact_path(self, test_repo):
@@ -1572,13 +1691,17 @@ class TestMetadataRepository:
         ]
 
     def test__bootstrap_finalize(self, test_repo, monkeypatch):
+        fake_key_dict = {"keyval": "foo"}
+        fake_key = pretend.stub(
+            to_dict=pretend.call_recorder(lambda: fake_key_dict), keyid="keyid"
+        )
         fake_role = pretend.stub(keyids=["keyid"])
         fake_root = pretend.stub(
             signatures=pretend.stub(clear=pretend.call_recorder(lambda: None)),
             sign=pretend.call_recorder(lambda *a, **kw: None),
             signed=pretend.stub(
                 roles={"timestamp": fake_role},
-                keys={"keyid": "online_key_id"},
+                keys={"keyid": fake_key},
             ),
         )
         test_repo._persist = pretend.call_recorder(lambda *a: None)
@@ -1603,11 +1726,12 @@ class TestMetadataRepository:
         assert fake_settings.get_fresh.calls == [
             pretend.call("CUSTOM_DELEGATED_ROLES")
         ]
-        assert test_repo._online_key == "online_key_id"
+        assert fake_key.to_dict.calls == [pretend.call()]
         assert test_repo._persist.calls == [
             pretend.call(fake_root, repository.Root.type)
         ]
         assert test_repo.write_repository_settings.calls == [
+            pretend.call("ONLINE_KEY", fake_key_dict),
             pretend.call("ROOT_SIGNING", None),
             pretend.call("BOOTSTRAP", "task_id"),
         ]
@@ -3555,7 +3679,7 @@ class TestMetadataRepository:
     def test__verify_new_root_signing(self, test_repo):
         fake_new_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
+                roles={"timestamp": pretend.stub(keyids=["k1"])},
                 version=2,
                 type=repository.Root.type,
             ),
@@ -3563,7 +3687,7 @@ class TestMetadataRepository:
         )
         fake_old_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
+                roles={"timestamp": pretend.stub(keyids=["k1"])},
                 version=1,
             ),
             verify_delegate=pretend.call_recorder(lambda *a: None),
@@ -3585,7 +3709,7 @@ class TestMetadataRepository:
     ):
         fake_new_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
+                roles={"timestamp": pretend.stub(keyids=["k1"])},
                 version=2,
                 type=repository.Root.type,
             ),
@@ -3595,7 +3719,7 @@ class TestMetadataRepository:
         )
         fake_old_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
+                roles={"timestamp": pretend.stub(keyids=["k1"])},
                 version=1,
             ),
             verify_delegate=pretend.call_recorder(lambda *a: None),
@@ -3610,14 +3734,14 @@ class TestMetadataRepository:
     def test__verify_new_root_signing_bad_version(self, test_repo):
         fake_new_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
+                roles={"timestamp": pretend.stub(keyids=["k1"])},
                 version=4,
                 type=repository.Root.type,
             ),
         )
         fake_old_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
+                roles={"timestamp": pretend.stub(keyids=["k1"])},
                 version=1,
             ),
         )
@@ -3631,14 +3755,14 @@ class TestMetadataRepository:
     def test__verify_new_root_signing_bad_type(self, test_repo):
         fake_new_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
+                roles={"timestamp": pretend.stub(keyids=["k1"])},
                 version=2,
                 type=repository.Snapshot.type,
             ),
         )
         fake_old_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"root": pretend.stub(keyids={"k1": "v1"})},
+                roles={"root": pretend.stub(keyids=["k1"])},
                 version=1,
             ),
         )
@@ -3853,13 +3977,13 @@ class TestMetadataRepository:
     def test__root_metadata_update(self, test_repo, mocked_datetime):
         fake_new_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
+                roles={"timestamp": pretend.stub(keyids=["k1"])},
                 version=2,
             )
         )
         fake_old_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
+                roles={"timestamp": pretend.stub(keyids=["k1"])},
                 version=1,
             )
         )
@@ -3898,13 +4022,13 @@ class TestMetadataRepository:
     ):
         fake_new_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
+                roles={"timestamp": pretend.stub(keyids=["k1"])},
                 version=2,
             )
         )
         fake_old_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
+                roles={"timestamp": pretend.stub(keyids=["k1"])},
                 version=1,
             )
         )
@@ -3945,13 +4069,13 @@ class TestMetadataRepository:
     ):
         fake_new_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
+                roles={"timestamp": pretend.stub(keyids=["k1"])},
                 version=2,
             )
         )
         fake_old_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
+                roles={"timestamp": pretend.stub(keyids=["k1"])},
                 version=1,
             )
         )
@@ -3977,18 +4101,37 @@ class TestMetadataRepository:
         ]
 
     def test__root_metadata_update_online_key(
-        self, test_repo, mocked_datetime
+        self, test_repo, mocked_datetime, monkeypatch
     ):
+        fake_key_dict = {"keyval": "foo", "keyid": "old_key_id"}
+        fake_settings = pretend.stub(
+            get_fresh=pretend.call_recorder(lambda a: copy(fake_key_dict))
+        )
+        monkeypatch.setattr(
+            repository,
+            "get_repository_settings",
+            lambda *a, **kw: fake_settings,
+        )
+        old_key = pretend.stub(keyid="old_key_id")
+        fake_key_obj = pretend.stub(
+            from_dict=pretend.call_recorder(lambda *a: old_key)
+        )
+        monkeypatch.setattr(f"{REPOSITORY_PATH}.Key", fake_key_obj)
+        new_key_dict = {"keyval": "bar"}
+        new_key = pretend.stub(
+            to_dict=pretend.call_recorder(lambda: new_key_dict),
+            keyid="new_key_id",
+        )
         fake_new_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids=["keyid2"])},
-                keys={"keyid2": "key2"},
+                roles={"timestamp": pretend.stub(keyids=["new_key_id"])},
+                keys={"new_key_id": new_key},
                 version=2,
             )
         )
         fake_old_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids=["keyid1"])},
+                roles={"timestamp": pretend.stub(keyids=["old_key_id"])},
                 version=1,
             )
         )
@@ -4010,7 +4153,9 @@ class TestMetadataRepository:
         test_repo._run_online_roles_bump = pretend.call_recorder(
             lambda **kw: None
         )
-        test_repo._online_key = "key1"
+        test_repo.write_repository_settings = pretend.call_recorder(
+            lambda *a: None
+        )
 
         result = test_repo._root_metadata_update(fake_new_root_md)
 
@@ -4036,23 +4181,30 @@ class TestMetadataRepository:
         assert test_repo._persist.calls == [
             pretend.call(fake_new_root_md, repository.Root.type)
         ]
+        assert fake_settings.get_fresh.calls == [pretend.call("ONLINE_KEY")]
+        assert fake_key_obj.from_dict.calls == [
+            pretend.call(fake_key_dict.pop("keyid"), fake_key_dict)
+        ]
+        assert new_key.to_dict.calls == [pretend.call()]
         assert test_repo._run_online_roles_bump.calls == [
             pretend.call(force=True)
         ]
-        assert test_repo._online_key == "key2"
+        assert test_repo.write_repository_settings.calls == [
+            pretend.call("ONLINE_KEY", new_key_dict)
+        ]
 
     def test__root_metadata_update_online_key_lock_timeout(
         self, monkeypatch, test_repo
     ):
         fake_new_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
+                roles={"timestamp": pretend.stub(keyids=["k1"])},
                 version=2,
             )
         )
         fake_old_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k2": "v2"})},
+                roles={"timestamp": pretend.stub(keyids=["k2"])},
                 version=1,
             )
         )
@@ -4090,18 +4242,41 @@ class TestMetadataRepository:
         ]
 
     def test_root_metadata_update_finalize_run_onlines_bump_fails(
-        self, test_repo
+        self, test_repo, monkeypatch
     ) -> None:
+        fake_key_dict = {"keyval": "foo", "keyid": "old_key_id"}
+        fake_settings = pretend.stub(
+            get_fresh=pretend.call_recorder(lambda a: copy(fake_key_dict))
+        )
+        monkeypatch.setattr(
+            repository,
+            "get_repository_settings",
+            lambda *a, **kw: fake_settings,
+        )
+        old_key_dict = {"keyval": "old_bar"}
+        old_key = pretend.stub(
+            to_dict=pretend.call_recorder(lambda: old_key_dict),
+            keyid="old_key_id",
+        )
+        fake_key_obj = pretend.stub(
+            from_dict=pretend.call_recorder(lambda *a: old_key)
+        )
+        monkeypatch.setattr(f"{REPOSITORY_PATH}.Key", fake_key_obj)
+        new_key_dict = {"keyval": "old_bar"}
+        new_key = pretend.stub(
+            to_dict=pretend.call_recorder(lambda: new_key_dict),
+            keyid="new_key_id",
+        )
         fake_new_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids=["keyid2"])},
-                keys={"keyid2": "key2"},
+                roles={"timestamp": pretend.stub(keyids=["new_key_id"])},
+                keys={"new_key_id": new_key},
                 version=2,
             )
         )
         fake_old_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids=["keyid1"])},
+                roles={"timestamp": pretend.stub(keyids=["old_key_id"])},
                 version=1,
             )
         )
@@ -4114,7 +4289,9 @@ class TestMetadataRepository:
             lock=pretend.call_recorder(mocked_lock),
         )
         test_repo._run_online_roles_bump = pretend.raiser(ValueError("Bad"))
-        test_repo._online_key = "key1"
+        test_repo.write_repository_settings = pretend.call_recorder(
+            lambda *a: None
+        )
 
         with pytest.raises(ValueError):
             test_repo._root_metadata_update_finalize(
@@ -4124,8 +4301,18 @@ class TestMetadataRepository:
         assert test_repo._redis.lock.calls == [
             pretend.call(repository.LOCK_TARGETS, timeout=60.0)
         ]
-        # Assert that we have recovered to the previous online key
-        assert test_repo._online_key == "key1"
+        assert fake_settings.get_fresh.calls == [pretend.call("ONLINE_KEY")]
+        assert fake_key_obj.from_dict.calls == [
+            pretend.call(fake_key_dict.pop("keyid"), fake_key_dict)
+        ]
+        assert new_key.to_dict.calls == [pretend.call()]
+        # Assert that we have recovered to the previous online key.
+        # These calls assert that the online key is set to current online key.
+        assert old_key.to_dict.calls == [pretend.call()]
+        assert test_repo.write_repository_settings.calls == [
+            pretend.call("ONLINE_KEY", new_key_dict),
+            pretend.call("ONLINE_KEY", old_key_dict),
+        ]
 
     def test_metadata_update(self, monkeypatch, test_repo):
         fake_settings = pretend.stub(
@@ -4645,7 +4832,7 @@ class TestMetadataRepository:
             pretend.call("ROOT_SIGNING", "fake_metadata")
         ]
 
-    def test_sign_metadata__update__bad_role_type(
+    def test_sign_metadata_update_bad_role_type(
         self, test_repo, monkeypatch, mocked_datetime
     ):
         fake_signature = pretend.stub(keyid="fake")
@@ -4664,7 +4851,7 @@ class TestMetadataRepository:
             "details": None,
         }
 
-    def test_sign_metadata__update__invalid_signature(
+    def test_sign_metadata_update_invalid_signature(
         self,
         test_repo,
         monkeypatch,
@@ -4730,7 +4917,7 @@ class TestMetadataRepository:
             pretend.call({"metadata": "fake"})
         ]
 
-    def test_sign_metadata__update__invalid_threshold__trusted_and_new(
+    def test_sign_metadata_update_invalid_threshold_trusted_and_new(
         self,
         test_repo,
         monkeypatch,
@@ -4796,7 +4983,7 @@ class TestMetadataRepository:
             },
         }
 
-    def test_sign_metadata__update__invalid_threshold__trusted(
+    def test_sign_metadata_update_invalid_threshold_trusted(
         self,
         test_repo,
         monkeypatch,
@@ -4862,7 +5049,7 @@ class TestMetadataRepository:
             },
         }
 
-    def test_sign_metadata__update__invalid_threshold__new(
+    def test_sign_metadata_update_invalid_threshold_new(
         self,
         test_repo,
         monkeypatch,
@@ -4930,7 +5117,7 @@ class TestMetadataRepository:
             },
         }
 
-    def test_sign_metadata__update__valid_threshold(
+    def test_sign_metadata_update_valid_threshold(
         self,
         test_repo,
         monkeypatch,
@@ -4979,6 +5166,9 @@ class TestMetadataRepository:
         test_repo._validate_threshold = pretend.call_recorder(
             lambda *a: next(fake_threshold_result)
         )
+        test_repo._root_metadata_update_finalize = pretend.call_recorder(
+            lambda *a: None
+        )
 
         # Call sign_metadata with fake payload
         # All deserialization and validation is mocked
@@ -4995,6 +5185,9 @@ class TestMetadataRepository:
                 "update": "Metadata update finished",
             },
         }
+        assert test_repo._root_metadata_update_finalize.calls == [
+            pretend.call(fake_trusted_root, fake_new_root)
+        ]
 
     def test_delete_sign_metadata_bootstrap_signing_state(
         self, test_repo, monkeypatch, mocked_datetime

--- a/tests/unit/tuf_repository_service_worker/test_repository.py
+++ b/tests/unit/tuf_repository_service_worker/test_repository.py
@@ -3443,7 +3443,7 @@ class TestMetadataRepository:
             pretend.call(repository.LOCK_TARGETS, timeout=60)
         ]
 
-    def test__trusted_root_update(self, test_repo):
+    def test__verify_new_root_signing(self, test_repo):
         fake_new_root_md = pretend.stub(
             signed=pretend.stub(
                 roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
@@ -3460,7 +3460,7 @@ class TestMetadataRepository:
             verify_delegate=pretend.call_recorder(lambda *a: None),
         )
 
-        result = test_repo._trusted_root_update(
+        result = test_repo._verify_new_root_signing(
             fake_old_root_md, fake_new_root_md
         )
         assert result is None
@@ -3471,7 +3471,7 @@ class TestMetadataRepository:
             pretend.call(repository.Root.type, fake_new_root_md)
         ]
 
-    def test__trusted_root_update_fail_current_verify_delegate(
+    def test__verify_new_root_signing_fail_current_verify_delegate(
         self, test_repo
     ):
         fake_new_root_md = pretend.stub(
@@ -3493,10 +3493,12 @@ class TestMetadataRepository:
         )
 
         with pytest.raises(TypeError) as err:
-            test_repo._trusted_root_update(fake_old_root_md, fake_new_root_md)
+            test_repo._verify_new_root_signing(
+                fake_old_root_md, fake_new_root_md
+            )
         assert "Call is valid only on delegator metadata" in str(err)
 
-    def test__trusted_root_update_bad_version(self, test_repo):
+    def test__verify_new_root_signing_bad_version(self, test_repo):
         fake_new_root_md = pretend.stub(
             signed=pretend.stub(
                 roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
@@ -3512,10 +3514,12 @@ class TestMetadataRepository:
         )
 
         with pytest.raises(repository.BadVersionNumberError) as err:
-            test_repo._trusted_root_update(fake_old_root_md, fake_new_root_md)
+            test_repo._verify_new_root_signing(
+                fake_old_root_md, fake_new_root_md
+            )
         assert "Expected root version 2 instead got version 4" in str(err)
 
-    def test__trusted_root_update_bad_type(self, test_repo):
+    def test__verify_new_root_signing_bad_type(self, test_repo):
         fake_new_root_md = pretend.stub(
             signed=pretend.stub(
                 roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
@@ -3531,7 +3535,9 @@ class TestMetadataRepository:
         )
 
         with pytest.raises(repository.RepositoryError) as err:
-            test_repo._trusted_root_update(fake_old_root_md, fake_new_root_md)
+            test_repo._verify_new_root_signing(
+                fake_old_root_md, fake_new_root_md
+            )
         assert "Expected 'root', got 'snapshot'" in str(err)
 
     def test_run_force_online_metadata_update_targets_and_bins(
@@ -3751,7 +3757,9 @@ class TestMetadataRepository:
         test_repo._storage_backend.get = pretend.call_recorder(
             lambda *a: fake_old_root_md
         )
-        test_repo._trusted_root_update = pretend.call_recorder(lambda *a: None)
+        test_repo._verify_new_root_signing = pretend.call_recorder(
+            lambda *a: None
+        )
         test_repo._persist = pretend.call_recorder(lambda *a: None)
 
         result = test_repo._root_metadata_update(fake_new_root_md)
@@ -3769,7 +3777,7 @@ class TestMetadataRepository:
         assert test_repo._storage_backend.get.calls == [
             pretend.call(repository.Root.type)
         ]
-        assert test_repo._trusted_root_update.calls == [
+        assert test_repo._verify_new_root_signing.calls == [
             pretend.call(fake_old_root_md, fake_new_root_md)
         ]
         assert test_repo._persist.calls == [
@@ -3794,7 +3802,7 @@ class TestMetadataRepository:
         test_repo._storage_backend.get = pretend.call_recorder(
             lambda *a: fake_old_root_md
         )
-        test_repo._trusted_root_update = pretend.raiser(
+        test_repo._verify_new_root_signing = pretend.raiser(
             repository.UnsignedMetadataError()
         )
 
@@ -3841,7 +3849,7 @@ class TestMetadataRepository:
         test_repo._storage_backend.get = pretend.call_recorder(
             lambda *a: fake_old_root_md
         )
-        test_repo._trusted_root_update = pretend.raiser(
+        test_repo._verify_new_root_signing = pretend.raiser(
             repository.BadVersionNumberError("Version v3 instead v2")
         )
 
@@ -3878,7 +3886,9 @@ class TestMetadataRepository:
         test_repo._storage_backend.get = pretend.call_recorder(
             lambda *a: fake_old_root_md
         )
-        test_repo._trusted_root_update = pretend.call_recorder(lambda *a: None)
+        test_repo._verify_new_root_signing = pretend.call_recorder(
+            lambda *a: None
+        )
 
         @contextmanager
         def mocked_lock(lock, timeout):
@@ -3908,7 +3918,7 @@ class TestMetadataRepository:
         assert test_repo._storage_backend.get.calls == [
             pretend.call(repository.Root.type)
         ]
-        assert test_repo._trusted_root_update.calls == [
+        assert test_repo._verify_new_root_signing.calls == [
             pretend.call(fake_old_root_md, fake_new_root_md)
         ]
         assert test_repo._redis.lock.calls == [
@@ -3948,7 +3958,9 @@ class TestMetadataRepository:
         test_repo._storage_backend.get = pretend.call_recorder(
             lambda *a: fake_old_root_md
         )
-        test_repo._trusted_root_update = pretend.call_recorder(lambda *a: None)
+        test_repo._verify_new_root_signing = pretend.call_recorder(
+            lambda *a: None
+        )
 
         @contextmanager
         def mocked_lock(lock, timeout):
@@ -3964,7 +3976,7 @@ class TestMetadataRepository:
         assert test_repo._storage_backend.get.calls == [
             pretend.call(repository.Root.type)
         ]
-        assert test_repo._trusted_root_update.calls == [
+        assert test_repo._verify_new_root_signing.calls == [
             pretend.call(fake_old_root_md, fake_new_root_md)
         ]
 

--- a/tests/unit/tuf_repository_service_worker/test_repository.py
+++ b/tests/unit/tuf_repository_service_worker/test_repository.py
@@ -191,17 +191,9 @@ class TestMetadataRepository:
         assert "No permission /run/secrets/*" == caplog.messages[0]
 
     def test__sign(self, test_repo):
-        fake_role = pretend.stub(keyids=["keyid_1"])
+        test_repo._online_key = "onlin_key"
         fake_md = pretend.stub(
-            signatures=pretend.stub(clear=pretend.call_recorder(lambda: None)),
             sign=pretend.call_recorder(lambda *a, **kw: None),
-            signed=pretend.stub(
-                roles={"timestamp": fake_role},
-                keys={"keyid_1": {}},
-            ),
-        )
-        test_repo._storage_backend = pretend.stub(
-            get=pretend.call_recorder(lambda *a: fake_md)
         )
         test_repo._signer_store = pretend.stub(
             get=pretend.call_recorder(lambda *a: "key_signer_1")
@@ -210,12 +202,8 @@ class TestMetadataRepository:
         test_result = test_repo._sign(fake_md)
 
         assert test_result is None
-        assert test_repo._signer_store.get.calls == [pretend.call({})]
-        assert fake_md.signatures.clear.calls == [pretend.call()]
-        assert test_repo._storage_backend.get.calls == [pretend.call("root")]
-        assert fake_md.sign.calls == [
-            pretend.call("key_signer_1", append=True),
-        ]
+        assert test_repo._signer_store.get.calls == [pretend.call("onlin_key")]
+        assert fake_md.sign.calls == [pretend.call("key_signer_1")]
 
     def _test_helper_persist(
         self, test_repo, role, version, expected_file_name
@@ -1487,6 +1475,15 @@ class TestMetadataRepository:
         ]
 
     def test__bootstrap_finalize(self, test_repo, monkeypatch):
+        fake_role = pretend.stub(keyids=["keyid"])
+        fake_root = pretend.stub(
+            signatures=pretend.stub(clear=pretend.call_recorder(lambda: None)),
+            sign=pretend.call_recorder(lambda *a, **kw: None),
+            signed=pretend.stub(
+                roles={"timestamp": fake_role},
+                keys={"keyid": "online_key_id"},
+            ),
+        )
         test_repo._persist = pretend.call_recorder(lambda *a: None)
         test_repo.write_repository_settings = pretend.call_recorder(
             lambda *a: None
@@ -1503,21 +1500,22 @@ class TestMetadataRepository:
             "get_repository_settings",
             lambda *a, **kw: fake_settings,
         )
-        result = test_repo._bootstrap_finalize("fake_root", "task_id")
+        result = test_repo._bootstrap_finalize(fake_root, "task_id")
 
         assert result is None
         assert fake_settings.get_fresh.calls == [
             pretend.call("CUSTOM_DELEGATED_ROLES")
         ]
+        assert test_repo._online_key == "online_key_id"
         assert test_repo._persist.calls == [
-            pretend.call("fake_root", repository.Root.type)
+            pretend.call(fake_root, repository.Root.type)
         ]
         assert test_repo.write_repository_settings.calls == [
             pretend.call("ROOT_SIGNING", None),
             pretend.call("BOOTSTRAP", "task_id"),
         ]
         assert test_repo._bootstrap_online_roles.calls == [
-            pretend.call("fake_root", fake_delegated_roles)
+            pretend.call(fake_root, fake_delegated_roles)
         ]
 
     def test_bootstrap(self, monkeypatch, test_repo, mocked_datetime):
@@ -3866,13 +3864,14 @@ class TestMetadataRepository:
     ):
         fake_new_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k1": "v1"})},
+                roles={"timestamp": pretend.stub(keyids=["keyid2"])},
+                keys={"keyid2": "key2"},
                 version=2,
             )
         )
         fake_old_root_md = pretend.stub(
             signed=pretend.stub(
-                roles={"timestamp": pretend.stub(keyids={"k2": "v2"})},
+                roles={"timestamp": pretend.stub(keyids=["keyid1"])},
                 version=1,
             )
         )
@@ -3892,6 +3891,7 @@ class TestMetadataRepository:
         test_repo._run_online_roles_bump = pretend.call_recorder(
             lambda **kw: None
         )
+        test_repo._online_key = "key1"
 
         result = test_repo._root_metadata_update(fake_new_root_md)
 
@@ -3920,6 +3920,7 @@ class TestMetadataRepository:
         assert test_repo._run_online_roles_bump.calls == [
             pretend.call(force=True)
         ]
+        assert test_repo._online_key == "key2"
 
     def test__root_metadata_update_online_key_lock_timeout(
         self, monkeypatch, test_repo
@@ -3966,6 +3967,44 @@ class TestMetadataRepository:
         assert test_repo._trusted_root_update.calls == [
             pretend.call(fake_old_root_md, fake_new_root_md)
         ]
+
+    def test_root_metadata_update_finalize_run_onlines_bump_fails(
+        self, test_repo
+    ) -> None:
+        fake_new_root_md = pretend.stub(
+            signed=pretend.stub(
+                roles={"timestamp": pretend.stub(keyids=["keyid2"])},
+                keys={"keyid2": "key2"},
+                version=2,
+            )
+        )
+        fake_old_root_md = pretend.stub(
+            signed=pretend.stub(
+                roles={"timestamp": pretend.stub(keyids=["keyid1"])},
+                version=1,
+            )
+        )
+
+        @contextmanager
+        def mocked_lock(lock, timeout):
+            yield lock, timeout
+
+        test_repo._redis = pretend.stub(
+            lock=pretend.call_recorder(mocked_lock),
+        )
+        test_repo._run_online_roles_bump = pretend.raiser(ValueError("Bad"))
+        test_repo._online_key = "key1"
+
+        with pytest.raises(ValueError):
+            test_repo._root_metadata_update_finalize(
+                fake_old_root_md, fake_new_root_md
+            )
+
+        assert test_repo._redis.lock.calls == [
+            pretend.call(repository.LOCK_TARGETS, timeout=60.0)
+        ]
+        # Assert that we have recovered to the previous online key
+        assert test_repo._online_key == "key1"
 
     def test_metadata_update(self, monkeypatch, test_repo):
         fake_settings = pretend.stub(
@@ -4556,6 +4595,19 @@ class TestMetadataRepository:
             "error": "Invalid signature",
             "details": None,
         }
+        assert fake_settings.get_fresh.calls == [
+            pretend.call("ROOT_SIGNING"),
+            pretend.call("BOOTSTRAP"),
+        ]
+        assert repository.Signature.from_dict.calls == [
+            pretend.call(payload["signature"])
+        ]
+        assert test_repo._storage_backend.get.calls == [
+            pretend.call(Root.type)
+        ]
+        assert repository.Metadata.from_dict.calls == [
+            pretend.call({"metadata": "fake"})
+        ]
 
     def test_sign_metadata__update__invalid_threshold__trusted_and_new(
         self,

--- a/tests/unit/tuf_repository_service_worker/test_repository.py
+++ b/tests/unit/tuf_repository_service_worker/test_repository.py
@@ -871,6 +871,103 @@ class TestMetadataRepository:
             pretend.call(version=mocked_targets.signed.version),
         ]
 
+    def test__update_targets_delegations_key_bins(self, test_repo):
+        fake_online_key_stub = pretend.stub(keyid="new_online_key")
+        test_repo._online_key = fake_online_key_stub
+        fake_root = pretend.stub(
+            signed=pretend.stub(
+                roles={Targets.type: pretend.stub(keyids=["old_keyid"])}
+            )
+        )
+        test_repo._storage_backend.get = pretend.call_recorder(
+            lambda a: fake_root
+        )
+        fake_targets = pretend.stub(
+            signed=pretend.stub(
+                delegations=pretend.stub(succinct_roles=True),
+                revoke_key=pretend.call_recorder(lambda a: None),
+                add_key=pretend.call_recorder(lambda a: None),
+            )
+        )
+
+        result = test_repo._update_targets_delegations_key(fake_targets)
+        assert result is None
+        assert test_repo._storage_backend.get.calls == [
+            pretend.call(Root.type)
+        ]
+        assert fake_targets.signed.revoke_key.calls == [
+            pretend.call("old_keyid")
+        ]
+        assert fake_targets.signed.add_key.calls == [
+            pretend.call(fake_online_key_stub)
+        ]
+
+    def test__update_targets_delegations_key_custom_delegations(
+        self, test_repo
+    ):
+        fake_online_key_stub = pretend.stub(keyid="new_online_key")
+        test_repo._online_key = fake_online_key_stub
+        fake_root = pretend.stub(
+            signed=pretend.stub(
+                roles={Targets.type: pretend.stub(keyids=["old_keyid"])}
+            )
+        )
+        test_repo._storage_backend.get = pretend.call_recorder(
+            lambda a: fake_root
+        )
+        fake_role1 = pretend.stub(name="role1")
+        fake_role2 = pretend.stub(name="role2")
+        fake_targets = pretend.stub(
+            signed=pretend.stub(
+                delegations=pretend.stub(
+                    succinct_roles=None,
+                    roles=pretend.stub(
+                        values=pretend.call_recorder(
+                            lambda: [fake_role1, fake_role2]
+                        )
+                    ),
+                ),
+                revoke_key=pretend.call_recorder(lambda *a: None),
+                add_key=pretend.call_recorder(lambda *a: None),
+            )
+        )
+
+        result = test_repo._update_targets_delegations_key(fake_targets)
+        assert result is None
+        assert test_repo._storage_backend.get.calls == [
+            pretend.call(Root.type)
+        ]
+        assert fake_targets.signed.delegations.roles.values.calls == [
+            pretend.call(),
+        ]
+        assert fake_targets.signed.revoke_key.calls == [
+            pretend.call("old_keyid", "role1"),
+            pretend.call("old_keyid", "role2"),
+        ]
+        assert fake_targets.signed.add_key.calls == [
+            pretend.call(fake_online_key_stub, "role1"),
+            pretend.call(fake_online_key_stub, "role2"),
+        ]
+
+    def test__update_targets_delegations_online_key_not_changed(
+        self, test_repo
+    ):
+        test_repo._online_key = pretend.stub(keyid="online_key")
+        fake_root = pretend.stub(
+            signed=pretend.stub(
+                roles={Targets.type: pretend.stub(keyids=["online_key"])}
+            )
+        )
+        test_repo._storage_backend.get = pretend.call_recorder(
+            lambda a: fake_root
+        )
+
+        result = test_repo._update_targets_delegations_key("targets")
+        assert result is None
+        assert test_repo._storage_backend.get.calls == [
+            pretend.call(Root.type)
+        ]
+
     def test__get_role_for_artifact_path(self, test_repo):
         fake_targets = pretend.stub(
             signed=pretend.stub(
@@ -2951,6 +3048,9 @@ class TestMetadataRepository:
             "get_repository_settings",
             lambda *a, **kw: fake_settings,
         )
+        test_repo._update_targets_delegations_key = pretend.call_recorder(
+            lambda a: None
+        )
         test_repo._bump_and_persist = pretend.call_recorder(lambda *a: None)
         test_repo._update_snapshot = pretend.call_recorder(
             lambda **kw: "fake_snapshot"
@@ -2973,6 +3073,9 @@ class TestMetadataRepository:
             pretend.call("TARGETS_ONLINE_KEY"),
             pretend.call("TARGETS_ONLINE_KEY"),
             pretend.call("DELEGATED_ROLES_NAMES"),
+        ]
+        assert test_repo._update_targets_delegations_key.calls == [
+            pretend.call(fake_targets)
         ]
         assert test_repo._bump_and_persist.calls == [
             pretend.call(fake_targets, Targets.type),
@@ -3010,6 +3113,9 @@ class TestMetadataRepository:
             "get_repository_settings",
             lambda *a, **kw: fake_settings,
         )
+        test_repo._update_targets_delegations_key = pretend.call_recorder(
+            lambda a: None
+        )
         test_repo._bump_and_persist = pretend.call_recorder(lambda *a: None)
         test_repo._update_snapshot = pretend.call_recorder(
             lambda **kw: "fake_snapshot"
@@ -3026,6 +3132,9 @@ class TestMetadataRepository:
         test_repo._run_online_roles_bump(force=True)
         assert test_repo._storage_backend.get.calls == [
             pretend.call(Targets.type),
+        ]
+        assert test_repo._update_targets_delegations_key.calls == [
+            pretend.call(fake_targets)
         ]
         assert test_repo._bump_and_persist.calls == [
             pretend.call(fake_targets, Targets.type),

--- a/tests/unit/tuf_repository_service_worker/test_repository.py
+++ b/tests/unit/tuf_repository_service_worker/test_repository.py
@@ -26,7 +26,7 @@ from tuf.api.metadata import (
 from repository_service_tuf_worker import Dynaconf, repository
 from repository_service_tuf_worker.models import targets_schema
 
-REPOSITORY_PATH = f"repository_service_tuf_worker.repository"
+REPOSITORY_PATH = "repository_service_tuf_worker.repository"
 
 
 class TestRoles:
@@ -103,7 +103,6 @@ class TestMetadataRepository:
         ]
 
     def test_online_key_property_from_storage(self, test_repo, monkeypatch):
-        fake_key_dict = {"keyval": "foo", "keyid": "key_id"}
         fake_settings = pretend.stub(
             get_fresh=pretend.call_recorder(lambda a: None)
         )

